### ISSUE TITLE
Failing to install numpy 1.26.0 on aarch64 with pypy39

### ISSUE THREAD

**Issue by mipierre** [numpy/numpy#24768]
### Steps to reproduce:

1. Provision an aarch64 based machine. 
2. Using pypy39 install numpy 1.26.0

### Error message:

```shell
| 15.6 MB 48.4 MB/s 0:00:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [296 lines of output]
      + /home/REDACTED_USER/REDACTED/t-env/bin/pypy /tmp/pip-install-io32ek1a/numpy/vendored-meson/meson/meson.py setup /tmp/pip-install-io32ek1a/numpy /tmp/pip-install-io32ek1a/numpy/.mesonpy-t184evq8/build -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --native-file=/tmp/pip-install-io32ek1a/numpy/.mesonpy-t184evq8/build/meson-python-native-file.ini
      The Meson build system
      Version: 1.2.99
      Source dir: /tmp/pip-install-io32ek1a/numpy
      Build dir: /tmp/pip-install-io32ek1a/numpy/.mesonpy-t184evq8/build
      Build type: native build
      Project name: NumPy
      Project version: 1.26.0
      C compiler for the host machine: cc (gcc 9.4.0 "cc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0")
      C linker for the host machine: cc ld.bfd 2.34
      C++ compiler for the host machine: c++ (gcc 9.4.0 "c++ (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0")
      C++ linker for the host machine: c++ ld.bfd 2.34
      Cython compiler for the host machine: cython (cython 3.0.2)
      Host machine cpu family: aarch64
      Host machine cpu: aarch64
      Program python found: YES (/home/REDACTED_USER/REDACTED/t-env/bin/pypy)
      Run-time dependency python found: YES 3.9
      Has header "Python.h" with dependency python: YES
      Compiler for C supports arguments -fno-strict-aliasing: YES
      Message: During parsing cpu-dispatch: The following CPU features were ignored due to platform incompatibility or lack of support:
      "XOP FMA4"
      Test features "NEON NEON_FP16 NEON_VFPV4 ASIMD" : Supported
      Test features "ASIMDHP" : Supported
      Test features "ASIMDFHM" : Supported
      Configuring npy_cpu_dispatch_config.h using configuration
      Message:
      CPU Optimization Options
        baseline:
          Requested : min
          Enabled   : NEON NEON_FP16 NEON_VFPV4 ASIMD
        dispatch:
          Requested : max -xop -fma4
          Enabled   : ASIMDHP ASIMDFHM

      Library m found: YES
      Found pkg-config: /usr/bin/pkg-config (0.29.1)
      Run-time dependency openblas found: YES 0.3.8
      Checking if "CBLAS" with dependency openblas: links: YES
      Dependency openblas found: YES 0.3.8 (cached)
      Program _build_utils/process_src_template.py found: YES (/home/REDACTED_USER/REDACTED/t-env/bin/pypy /tmp/pip-install-io32ek1a/numpy/numpy/_build_utils/process_src_template.py)
      Program _build_utils/tempita.py found: YES (/tmp/pip-install-io32ek1a/numpy/numpy/_build_utils/tempita.py)
      Configuring __config__.py using configuration
      Checking for size of "short" : 2
      Checking for size of "int" : 4
      Checking for size of "long" : 8
      Checking for size of "long long" : 8
      Checking for size of "float" : 4
      Checking for size of "double" : 8
      Checking for size of "long double" : 16
      Checking for size of "off_t" : 8
      Checking for size of "Py_intptr_t" with dependency python: -1
      Checking for size of "PY_LONG_LONG" with dependency python: -1
      Has header "complex.h" : YES
      Checking for type "complex float" : YES
      Checking for size of "struct {float __x; float __y;}" : 8
      Checking for type "complex double" : YES
      Checking for size of "struct {double __x; double __y;}" : 16
      Checking for type "complex long double" : YES
      Checking for size of "struct {long double __x; long double __y;}" : 32
      Checking for function "sin" : YES
      Checking for function "cos" : YES
      Checking for function "tan" : YES
      Checking for function "sinh" : YES
      Checking for function "cosh" : YES
      Checking for function "tanh" : YES
      Checking for function "fabs" : YES
      Checking for function "floor" : YES
      Checking for function "ceil" : YES
      Checking for function "sqrt" : YES
      Checking for function "log10" : YES
      Checking for function "log" : YES
      Checking for function "exp" : YES
      Checking for function "asin" : YES
      Checking for function "acos" : YES
      Checking for function "atan" : YES
      Checking for function "fmod" : YES
      Checking for function "modf" : YES
      Checking for function "frexp" : YES
      Checking for function "ldexp" : YES
      Checking for function "expm1" : YES
      Checking for function "log1p" : YES
      Checking for function "acosh" : YES
      Checking for function "asinh" : YES
      Checking for function "atanh" : YES
      Checking for function "rint" : YES
      Checking for function "trunc" : YES
      Checking for function "exp2" : YES
      Checking for function "copysign" : YES
      Checking for function "nextafter" : YES
      Checking for function "strtoll" : YES
      Checking for function "strtoull" : YES
      Checking for function "cbrt" : YES
      Checking for function "log2" : YES
      Checking for function "pow" : YES
      Checking for function "hypot" : YES
      Checking for function "atan2" : YES
      Checking for function "csin" : YES
      Checking for function "csinh" : YES
      Checking for function "ccos" : YES
      Checking for function "ccosh" : YES
      Checking for function "ctan" : YES
      Checking for function "ctanh" : YES
      Checking for function "creal" : YES
      Checking for function "cimag" : YES
      Checking for function "conj" : YES
      Checking for function "cabs" : YES
      Checking for function "cabsf" : YES
      Checking for function "cabsl" : YES
      Checking for function "cacos" : YES
      Checking for function "cacosf" : YES
      Checking for function "cacosl" : YES
      Checking for function "cacosh" : YES
      Checking for function "cacoshf" : YES
      Checking for function "cacoshl" : YES
      Checking for function "carg" : YES
      Checking for function "cargf" : YES
      Checking for function "cargl" : YES
      Checking for function "casin" : YES
      Checking for function "casinf" : YES
      Checking for function "casinl" : YES
      Checking for function "casinh" : YES
      Checking for function "casinhf" : YES
      Checking for function "casinhl" : YES
      Checking for function "catan" : YES
      Checking for function "catanf" : YES
      Checking for function "catanl" : YES
      Checking for function "catanh" : YES
      Checking for function "catanhf" : YES
      Checking for function "catanhl" : YES
      Checking for function "cexp" : YES
      Checking for function "cexpf" : YES
      Checking for function "cexpl" : YES
      Checking for function "clog" : YES
      Checking for function "clogf" : YES
      Checking for function "clogl" : YES
      Checking for function "cpow" : YES
      Checking for function "cpowf" : YES
      Checking for function "cpowl" : YES
      Checking for function "csqrt" : YES
      Checking for function "csqrtf" : YES
      Checking for function "csqrtl" : YES
      Checking for function "csin" : YES (cached)
      Checking for function "csinf" : YES
      Checking for function "csinl" : YES
      Checking for function "csinh" : YES (cached)
      Checking for function "csinhf" : YES
      Checking for function "csinhl" : YES
      Checking for function "ccos" : YES (cached)
      Checking for function "ccosf" : YES
      Checking for function "ccosl" : YES
      Checking for function "ccosh" : YES (cached)
      Checking for function "ccoshf" : YES
      Checking for function "ccoshl" : YES
      Checking for function "ctan" : YES (cached)
      Checking for function "ctanf" : YES
      Checking for function "ctanl" : YES
      Checking for function "ctanh" : YES (cached)
      Checking for function "ctanhf" : YES
      Checking for function "ctanhl" : YES
      Checking for function "isfinite" : YES
      Header "Python.h" has symbol "isfinite" with dependency python: YES
      Checking for function "isinf" : YES
      Header "Python.h" has symbol "isinf" with dependency python: YES
      Checking for function "isnan" : YES
      Header "Python.h" has symbol "isnan" with dependency python: YES
      Checking for function "signbit" : YES
      Header "Python.h" has symbol "signbit" with dependency python: YES
      Checking for function "fallocate" : YES
      Header "Python.h" has symbol "HAVE_FTELLO" with dependency python: NO
      Checking for function "ftello" : YES
      Header "Python.h" has symbol "HAVE_FSEEKO" with dependency python: NO
      Checking for function "fseeko" : YES
      Checking for function "strtold_l" : NO
      Checking for function "strtold_l" : NO
      Checking for function "backtrace" : YES
      Checking for function "madvise" : YES
      Has header "features.h" : YES
      Has header "xlocale.h" : NO
      Has header "dlfcn.h" : YES
      Has header "execinfo.h" : YES
      Has header "libunwind.h" : NO
      Has header "sys/mman.h" : YES
      Compiler for C supports arguments -O3: YES
      Has header "endian.h" : YES
      Has header "sys/endian.h" : NO
      Header "inttypes.h" has symbol "PRIdPTR" : YES
      Compiler for C supports function attribute visibility:hidden: YES
      Configuring config.h using configuration
      Configuring _numpyconfig.h using configuration
      Configuring npymath.ini using configuration
      Configuring mlib.ini using configuration
      Generating multi-targets for "_umath_tests.dispatch.h"
        Enabled targets: ASIMDHP, baseline
      Generating multi-targets for "argfunc.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "simd_qsort.dispatch.h"
        Enabled targets:
      Generating multi-targets for "simd_qsort_16bit.dispatch.h"
        Enabled targets:
      Generating multi-targets for "loops_arithm_fp.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_arithmetic.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_comparison.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_exponent_log.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_hyperbolic.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_logical.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_minmax.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_modulo.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_trigonometric.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_umath_fp.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_unary.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_unary_fp.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_unary_fp_le.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_unary_complex.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "loops_autovec.dispatch.h"
        Enabled targets: baseline
      Generating multi-targets for "_simd.dispatch.h"
        Enabled targets: baseline
      Build targets in project: 60

      NumPy 1.26.0

        User defined options
          Native files: /tmp/pip-install-io32ek1a/numpy/.mesonpy-t184evq8/build/meson-python-native-file.ini
          buildtype   : release
          b_ndebug    : if-release
          b_vscrt     : md

      Found ninja-1.11.1.git.kitware.jobserver-1 at /tmp/pip-build-env-04wuqyco/normal/bin/ninja
      + /tmp/pip-build-env-04wuqyco/normal/bin/ninja
      [1/299] Generating 'numpy/core/libnpymath.a.p/ieee754.c'
      [2/299] Generating numpy/__init__.cython-30.pxd with a custom command
      [3/299] Generating numpy/core/npy_math_internal.h with a custom command
      [4/299] Generating numpy/core/_umath_doc_generated with a custom command
      [5/299] Generating numpy/__init__.py with a custom command
      [6/299] Generating numpy/core/__umath_generated with a custom command
      [7/299] Generating 'numpy/core/libnpymath.a.p/npy_math_complex.c'
      [8/299] Generating numpy/__init__.pxd with a custom command
      [9/299] Compiling C object numpy/core/libnpymath.a.p/meson-generated_ieee754.c.o
      [10/299] Compiling C object numpy/core/libnpymath.a.p/meson-generated_npy_math_complex.c.o
      [11/299] Compiling C object numpy/core/libnpymath.a.p/src_npymath_npy_math.c.o
      [12/299] Generating 'numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/_multiarray_tests.c'
      [13/299] Generating 'numpy/core/_umath_tests.pypy39-pp73-aarch64-linux-gnu.so.p/_umath_tests.c'
      [14/299] Generating 'numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/templ_common.h'
      [15/299] Generating 'numpy/core/libargfunc.dispatch.h_baseline.a.p/arraytypes.h'
      [16/299] Generating 'numpy/core/libargfunc.dispatch.h_baseline.a.p/npy_sort.h'
      [17/299] Generating numpy/core/__ufunc_api with a custom command
      [18/299] Generating 'numpy/core/libloops_arithm_fp.dispatch.h_baseline.a.p/loops_utils.h'
      [19/299] Generating 'numpy/core/libloops_arithm_fp.dispatch.h_baseline.a.p/loops.h'
      [20/299] Generating 'numpy/core/libloops_arithm_fp.dispatch.h_baseline.a.p/loops_arithm_fp.dispatch.c'
      [21/299] Generating 'numpy/core/libargfunc.dispatch.h_baseline.a.p/argfunc.dispatch.c'
      [22/299] Generating 'numpy/core/libloops_arithmetic.dispatch.h_baseline.a.p/loops_utils.h'
      [23/299] Generating 'numpy/core/libloops_arithmetic.dispatch.h_baseline.a.p/loops.h'
      [24/299] Generating 'numpy/core/libloops_arithmetic.dispatch.h_baseline.a.p/loops_arithmetic.dispatch.c'
      [25/299] Generating numpy/core/__multiarray_api with a custom command
      [26/299] Generating 'numpy/core/libloops_comparison.dispatch.h_baseline.a.p/loops_utils.h'
      [27/299] Generating 'numpy/core/libloops_comparison.dispatch.h_baseline.a.p/loops.h'
      [28/299] Generating 'numpy/core/libloops_exponent_log.dispatch.h_baseline.a.p/loops.h'
      [29/299] Compiling C object numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/meson-generated__multiarray_tests.c.o
      FAILED: numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/meson-generated__multiarray_tests.c.o
      cc -Inumpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p -Inumpy/core -I../../numpy/core -I../../numpy/core/src/multiarray -I../../numpy/core/src/npymath -Inumpy/core/include -I../../numpy/core/include -I../../numpy/core/src/common -I/opt/pypy39/include/pypy3.9 -I/tmp/pip-install-io32ek1a/numpy/.mesonpy-t184evq8/build/meson_cpu -fvisibility=default -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -std=c99 -O3 -fno-strict-aliasing -DNPY_HAVE_NEON_VFPV4 -DNPY_HAVE_NEON_FP16 -DNPY_HAVE_NEON -DNPY_HAVE_ASIMD -fPIC -DNPY_INTERNAL_BUILD -DHAVE_NPY_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -MD -MQ numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/meson-generated__multiarray_tests.c.o -MF numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/meson-generated__multiarray_tests.c.o.d -o numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/meson-generated__multiarray_tests.c.o -c numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/_multiarray_tests.c
      In file included from ../../numpy/core/src/multiarray/_multiarray_tests.c.src:11:
      ../../numpy/core/src/multiarray/common.h: In function ‘check_and_adjust_index’:
      ../../numpy/core/src/multiarray/common.h:113:35: error: expected ‘)’ before ‘NPY_INTP_FMT’
        113 |                          "index %"NPY_INTP_FMT" is out of bounds "
            |                                   ^~~~~~~~~~~~
            |                                   )
      ../../numpy/core/src/multiarray/common.h:118:35: error: expected ‘)’ before ‘NPY_INTP_FMT’
        118 |                          "index %"NPY_INTP_FMT" is out of bounds "
            |                                   ^~~~~~~~~~~~
            |                                   )
      ../../numpy/core/src/multiarray/_multiarray_tests.c.src: In function ‘get_c_wrapping_array’:
      ../../numpy/core/src/multiarray/_multiarray_tests.c.src:1031:39: error: ‘NPY_INTP’ undeclared (first use in this function); did you mean ‘NPY_INT’?
       1031 |     descr =  PyArray_DescrNewFromType(NPY_INTP);
            |                                       ^~~~~~~~
            |                                       NPY_INT
      ../../numpy/core/src/multiarray/_multiarray_tests.c.src:1031:39: note: each undeclared identifier is reported only once for each function it appears in
      [30/299] Generating 'numpy/core/libloops_comparison.dispatch.h_baseline.a.p/loops_comparison.dispatch.c'
      [31/299] Compiling C object numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/src_common_npy_argparse.c.o
      [32/299] Compiling C object numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/src_common_npy_hashtable.c.o
      [33/299] Compiling C object numpy/core/_multiarray_tests.pypy39-pp73-aarch64-linux-gnu.so.p/src_common_mem_overlap.c.o
      [34/299] Compiling C++ object numpy/core/libnpymath.a.p/src_npymath_halffloat.cpp.o
      ninja: build stopped: subcommand failed.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> numpy

note: This is an issue with the package mentioned above, not pip.
```
```


### Additional information:

I was also installed `libopenblas-dev` prior to the install. Possibly there is an additional requirement missing to build the latest version on aarch64?

**Comment by mattip** [comment]
What is the complete pypy version `pypy3.9 -V`? You should use pypy3.9-v7.3.12. It seems you are using a ubuntu20.04-based image, can you use a newer one?

**Comment by seberg** [comment]
The problem here are those two lines:
```
      Checking for size of "Py_intptr_t" with dependency python: -1
      Checking for size of "PY_LONG_LONG" with dependency python: -1
```
I would hope updating `pypy` fixes this, I am a bit surprised since NumPy had this requirement of `Py_intptr_t` to be defined since forever?  And it isn't clear to me that we ever had a fallback in place that might have kicked in before meson.

In either case, we need the follow up on the version to keep the issue open.

**Comment by rgommers** [comment]
We have working PyPy 3.9 wheels for the 1.26.4 and 2.0.0b1 releases, so I assume this has long since been fixed - closing.

**Comment by 913085** [comment]
Not work I am lots of try but not solve what can I'd

**Comment by mattip** [comment]
@913085 please open a new issue and tell us
- what version of python
- what did you try
- what is the error message
